### ISSUE TITLE
FIREBASE_CONFIG not set when Node.js 10 function deployed

### ISSUE THREAD

**Issue by Bob-Snyder** [firebase/firebase-functions#433]
### [REQUIRED] Environment info

**firebase-tools:** 6.8.0

**Platform:** Windows 10

### [REQUIRED] Test case
I had a set of functions built using Node.js version 6.  They were successfully deployed and running.

I took the following actions to upgrade to use a newer version of Node:
  -- Upgraded my Windows development system to Node.js version 10.15.3
  -- Upgraded to latest Firebase tools: `npm install -g firebase-tools`
  -- Upgraded Firebase SDK: `npm install firebase-functions@latest firebase-admin@latest --save`
  -- Edited the project's `package.json` file to include `"engines" : { "node" : "10" }`

With those changes, the `firebase deploy --only functions` command  completed without error.  However, immediately after deployment, an error log appeared in the Firebase console for each function: `Warning, FIREBASE_CONFIG environment variable is missing. Initializing firebase-admin will fail`.

Changing `package.json file` to specify version 8 instead of 10 resolved the problem:  `"engines" : { "node" : "8" }` .  Functions initialize and execute as expected.

### [REQUIRED] Steps to reproduce
Described in test case

### [REQUIRED] Expected behavior
Firebase tools should set FIREBASE_CONFIG to the appropriate default for the project.

### [REQUIRED] Actual behavior
FIREBASE_CONFIG is not set if Node version of 10 is specified in `package.json` file.

**Comment by Bob-Snyder** [comment]
@thechenky
When the function executes, the only logs produced are these two:
 
`Function execution started`
`Function execution took 392 ms, finished with status code: 401`

The global initialization for my functions is:

```
const functions = require('firebase-functions');
const admin = require('firebase-admin');
admin.initializeApp();

const db = admin.firestore();

const settings = {timestampsInSnapshots: true};
db.settings(settings);

```

Here's one of the functions.  As I noted above, when execution fails, none of the expected log messages appear; only the exit status of 401.

```
exports.httpReportMessage2 = functions.https.onCall((reportMessage, context) => {
  // User must be authenticated.
  if (!context.auth) {
    throw new functions.https.HttpsError('failed-precondition', 'Not authenticated');
  }
  console.log('uid=', context.auth.uid);

  const timeSent = new Date(reportMessage.timeSent);

  console.log('timeSent=', timeSent.toString());

  const elapsed = Date.now() - timeSent;

  if (elapsed > 5000) {
    console.error('Report latency=', elapsed);
  }

  const message = JSON.parse(reportMessage.body);

  const enqueTime = new Date(message.enqueTime);

  const timeInQueue = timeSent - enqueTime;
  if (timeInQueue < 15000) {
    console.log('Time in queue=', timeInQueue);
  } else {
    console.error('Time in queue=', timeInQueue);
  }

  if (validReportMessage(message)) {
    console.log('Valid report message');
    return processReportMessage(message).then(() => {
      return 'Okay';
    }).catch(error => {
      return 'Error';
    });
  } else {
    return 'Invalid report';
  }
});

```

**Comment by rilian** [comment]
i have the same issue, and it happens when code access the runtime configuration values set by `firebase functions:config:set`, in this example it is my `cors_origin` value:

```js
"use strict";

const cors      = require("cors");
const functions = require("firebase-functions");

exports.cors = cors({
  origin: functions.config().cors_origin   // <------ stacktrace shows this line
});
```

```
Detailed stack trace: TypeError: Cannot read property 'cors_origin' of undefined
```

meaning `functions.config()` returns `undefined`

everything works fine on node 8

probably firebase-functions package is outdated and we should wait for it's next release ?

**Comment by thechenky** [comment]
Thanks @rilian for that report as well - these issues are connected since both `functions.config()` and `FIREBASE_CONFIG` are populated by reading in the `.runtimeconfig.json` file, which the SDK can no longer find in the Node 10 runtime. 

While I have a pending SDK fix for this, upon further investigating, I think this is a bug in the functions backend, and should be fixed there (see https://github.com/firebase/firebase-functions/pull/434#issuecomment-490168973). We are working with the relevant teams to resolve these issues. Thanks for everyone's patience!

**Comment by rilian** [comment]
thank you @thechenky

sorry for unrelated question, but which of these `functions.config()` or `FIREBASE_CONFIG` is preferred to use across the application, once this will be fixed  ?

**Comment by thechenky** [comment]
@rilian they serve different purposes. `functions.config()` is used if you're storing some custom key value pairs. To set the keys, you would do `firebase functions:config:set some-service.key=value`. To use that value in your functions code, retrieve it as you did with `functions.config().some-service.key` (see these [docs](https://firebase.google.com/docs/functions/config-env) for more info).

`FIREBASE_CONFIG` is something that you shouldn't need to worry about - it is automatically populated with your project information and allows you to use the Firebase Admin SDK to talk to Firebase, RTDB, or Storage, for example:

```
const admin = require('firebase-admin');
// FIREBASE_CONFIG is used here so you can initialize without params here:
admin.initializeApp();

// in your function:
exports.addMessage = functions.https.onRequest((req, res) => {
  // ...
  // FIREBASE_CONFIG helps you be able to use admin like this, by automatically 
  // knowing which Database URL you want to talk to
  admin.database().ref('/messages').push({text: 'hi'});
}
```
These [docs](https://firebase.google.com/docs/functions/config-env#automatically_populated_environment_variables) provide more info on what `FIREBASE_CONFIG` looks like.

Does that help?

**Comment by peterpeterparker** [comment]
Same here, with Node v8 all fine, when I upgrade to Node v10 I get the following error which I provide with the stack trace I found in the Firebase console in case that would help:

```
Error: process.env.GCLOUD_PROJECT is not set.
    at DocumentBuilder (/srv/functions/node_modules/firebase-functions/lib/providers/firestore.js:99:23)
    at cloudFunctionNewSignature (/srv/functions/node_modules/firebase-functions/lib/cloud-functions.js:102:13)
    at cloudFunction (/srv/functions/node_modules/firebase-functions/lib/cloud-functions.js:151:20)
    at Promise.resolve.then (/srv/node_modules/@google-cloud/functions-framework/build/src/invoker.js:330:28)
    at process._tickCallback (internal/process/next_tick.js:68:7)
```

**Comment by MaximBazarov** [comment]
I have the same on prod with Node 8 as well 
```
Error: process.env.GCLOUD_PROJECT is not set.
    at DocumentBuilder (/srv/functions/node_modules/firebase-functions/lib/providers/firestore.js:99:23)
    at cloudFunctionNewSignature (/srv/functions/node_modules/firebase-functions/lib/cloud-functions.js:102:13)
    at cloudFunction (/srv/functions/node_modules/firebase-functions/lib/cloud-functions.js:144:20)
    at Promise.resolve.then (/srv/node_modules/@google-cloud/functions-framework/build/src/invoker.js:330:28)
    at process._tickCallback (internal/process/next_tick.js:68:7)
```

**Comment by thechenky** [comment]
Quick update: we have identified and implemented the fix for unpopulated `FIREBASE_CONFIG` and `functions.config()` values. This should roll out to production sometime next week.

@MaximBazarov @peterpeterparker your issues are likely different and relating to `GCLOUD_PROJECT` environment variable missing from the Node 10 runtime (although it should work fine with Node 8). Please open a new issue about this.

**Comment by peterpeterparker** [comment]
@thechenky thx. I opened a new issue, let me know if you need further information

https://github.com/firebase/firebase-functions/issues/437

**Comment by IljaDaderko** [comment]
Just to confirm, I see same error on https callable functions after upgrading to node 10

> Warning, FIREBASE_CONFIG environment variable is missing. Initializing firebase-admin will fail

**Comment by jaschaio** [comment]
Seeing the same error:

`Function failed on loading user code. Error message: Warning, FIREBASE_CONFIG environment variable is missing. Initializing firebase-admin will fail`

Calls to `functions.config()` return undefined and thus crash the function deployment.

Within the functions emulator I get as well the error `Default "firebase-admin" instance created!` when calling `firebaseAdmin.initializeApp();` without parameters (which worked fine within node v8)

**Comment by jpreynat** [comment]
We are experiencing the exact same issue after migrating to `firebase-tools@^6.10.0` and the Node 10 runtime:
```
Warning, FIREBASE_CONFIG environment variable is missing. Initializing firebase-admin will fail
Provided module can't be loaded.
Is there a syntax error in your code?
```

The detailed stack trace indicates that reading the value from `functions.config()` returned `undefined`.

**Comment by idanen** [comment]
> Quick update: we have identified and implemented the fix for unpopulated `FIREBASE_CONFIG` and `functions.config()` values. This should roll out to production sometime next week.

Will you close this issue when it rolls out?

**Comment by yoavniran** [comment]
wasted an hour trying to figure out why my new function's config was always undefined after i deployed it. Searching google for "firebase function config empty" or "firebase function config undefined" didnt return anything useful. 

found this issue only after trying my last resort of reverting the engine from 10 to 8... :(

**Comment by thechenky** [comment]
The fix has been rolled out to production. `FIREBASE_CONFIG` and `functions.config()` should now work properly in Node 10 runtime. Please let me know if anyone is still seeing issues. Thanks!

**Comment by henrylearn2rock** [comment]
C:\Users\...\> firebase deploy --only functions

```
=== Deploying to 'xxxxxxx'...

i  deploying functions
i  functions: ensuring necessary APIs are enabled...
+  functions: all necessary APIs are enabled
i  functions: preparing functions directory for uploading...

Error: package.json in functions directory has an engines field which is unsupported. The only valid choices are: {"node": "8"} and {"node": "6"}.

```
C:\Users\...\> firebase -V
6.5.0
C:\Users\...\> node --version
v10.15.3

**Comment by thechenky** [comment]
Update firebase-tools

**Comment by loicmarie** [comment]
Is the fix working with the firebase functions emulator ? Cause I have the same error with the same versions
`ERROR Error: Unauthenticated
...`

**Comment by skhaz** [comment]
@thechenky still not working.

**Comment by thechenky** [comment]
@skhaz what exactly is not working? I'm going to need more information than that...

**Comment by skhaz** [comment]
@thechenky Sorry for the lack of information.

I tried a deploy using NodeJs 10 and got: 

`Error: process.env.GCLOUD_PROJECT is not set. at DocumentBuilder...`

**Comment by thechenky** [comment]
This is a separate issue: https://github.com/firebase/firebase-functions/issues/437. Please follow that bug.

**Comment by skhaz** [comment]
Ok. Thanks.

**Comment by FredyC** [comment]
For anyone stumbling about this issue where deployment fails to read from `functions.config()`, make sure you are not calling it outside the scope of function, ie. just above the function. I am pretty sure it must be some recent change because this was working for me for a while now.

**Comment by wsplou** [comment]
Have followed the guide https://medium.com/firebase-developers/migrate-your-firebase-cloud-functions-to-node-js-10-d9c677933ee to upgrade Node.js 8 to 10

I have the following error when deploy with `"engines": {
    "node": "10"
  }`, change it back to `"node": "8"` is deployed fine without any error.

`!  functions[test(us-central1)]: Deployment error.
Build failed: fatal: not a git repository (or any parent up to mount point /)
Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).
npm ERR! code ELIFECYCLE
npm ERR! errno 128`

**Comment by thaomd-2569** [comment]
@thechenky 
I'm using logger from firebase-functions and give this warning.Please help me fix it
![image](https://user-images.githubusercontent.com/86293405/137660825-d041ad0f-5473-4def-abe9-1965a45400c0.png)

**Comment by bolds07** [comment]
For those as me who had this annoying problem here comes the solution...

For some stupid reason firebase forces you to have this environment variables set EVEN IF YOU DON'T USE THEM (what was my case). to fix it you have 2 options

1- open google cloud functions -> click the function name -> edit -> add runtime variable -> them you have to manually inform the values of these two variables... THIS MUST BE DONE FOR EACH OF YOUR FUNCTIONS

2- you can copy paste the following code as the FIRST LINE of each of your firebase functions

```
process.env.GCLOUD_PROJECT = "YOURPROJECT ID";
process.env.FIREBASE_CONFIG = '{"projectId":"PROJECTID","databaseURL":"DATABASEURL","storageBucket":"BUCKET"}';
```

### ATTENTION!!!

the `FIREBASE_CONFIG ` isn't an object it is A STRING in json format so pay attention while setting it (these guys from firebase somethings do the dumbest things)

after that you stop getting this error

**Comment by taeold** [comment]
Another way I recommend would be to re-deploy your functions via `firebase deploy --only functions`. As long as you are using somewhat recent version of the Firebase CLI, it should take care of setting up the necessary environment variables in order for the Functions SDK to operate correctly.

`FIREBASE_CONFIG` is a string because it's an environment variable. What else, other than a string, can an environment variable be?
